### PR TITLE
fix(core): globally prohibit massive action if necessary

### PR DESCRIPTION
**Pull request by stonebuzz**
Even if  system user ```glpi-user``` can't be deleted from his form

```MassiveAction``` are available (with ```trashbin``` and ```purge```)

image

image

Add ```CommonDBTM``` function to prohibit massive action (```bulk``` or ```single```) if necessary

image

image

| Q             | A
| ------------- | ---
| Bug fix?      | yes/no
| New feature?  | yes/no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

**Comment by trasher** [comment]
With that change, massive actions list will differs if we sleect 10 users including "system user" or not, right? I'm not sure it's an iunderstandable behavior.

Maybe a better solution would just be to display a message telling that user cannot be removed, and let the MA do it works on all other.

**Comment by trasher** [comment]
> With that change, massive actions list will differs if we sleect 10 users including "system user" or not, right? I'm not sure it's an iunderstandable behavior.
> 
> Maybe a better solution would just be to display a message telling that user cannot be removed, and let the MA do it works on all other.

Nevermind, I misunderstood the change

File: src/User.php [hunk 1]
```diff
@@ -197,6 +202,13 @@ public function isEntityAssign()
     }
 
 
+    public static function isMassiveActionAllowed(int $items_id): bool
+    {
+        $config = Config::getConfigurationValues('core', ['system_user']);
```
Comment (cedric-anne) [inline src/User.php:207]: On huge lists, it will do many DB requests. Is this value available in `$CFG_GLPI`?

Comment (stonebuzz) [reply src/User.php]: yes =) i added a commit related to this change

File: src/User.php [hunk 2]
```diff
@@ -174,6 +174,11 @@ public function canDeleteItem()
             return false;
         }
 
+        //prevent delete / purge from API
+        if (!self::isMassiveActionAllowed($this->fields['id'])) {
+            return false;
+        }
```
Comment (cedric-anne) [inline src/User.php:180]: Maybe the whole PR can be replaced by
```suggestion
        global $CFG_GLPI;
        if ($this->fields['id'] == $CFG_GLPI['system_user']) {
            return false;
        }
```

Comment (AdrienClairembault) [reply src/User.php:180]: It may confuse the users, I think @stonebuzz solution where the checkbox is disabled is clearer.

Comment (stonebuzz) [reply src/User.php:180]: This part can be effectively simplified.

But is only useful in the case of a deletion from the

Comment (stonebuzz) [reply src/User.php:180]: Last commit use direct check instead of call ```isMassiveActionAllowed```

File: src/CommonDBTM.php [hunk 3]
```diff
@@ -6151,6 +6151,11 @@ public function getMassiveActionsForItem(): MassiveAction
         return new MassiveAction($params, [], 'initial', $this->getID());
     }
```
Comment (cedric-anne) [inline src/CommonDBTM.php:6153]:
```suggestion
    /**
     * Check whether actions are allowed for given item.
     */
```

File: src/CommonDBTM.php [hunk 4]
```diff
@@ -6150,6 +6150,13 @@ public function getMassiveActionsForItem(): MassiveAction
 
         return new MassiveAction($params, [], 'initial', $this->getID());
     }
+    /**
```
Comment (cedric-anne) [inline src/CommonDBTM.php:6153]:
```suggestion

    /**
```

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**